main.js

```diff
@@ -1,118 +1,78 @@
-let {Cu} = require("chrome");
-Cu.import("resource://gre/modules/Services.jsm");
-let adb = require("./adb");
-let fastboot = require("./fastboot");
-const events = require("sdk/event/core");
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+"use strict";
 
-const {devtools} = Cu.import("resource://gre/modules/devtools/Loader.jsm", {});
+const { Cu } = require("chrome");
+const { devtools } =
+  Cu.import("resource://gre/modules/devtools/Loader.jsm", {});
 const devtoolsRequire = devtools.require;
-const {ConnectionManager} = devtoolsRequire("devtools/client/connection-manager");
-let {Devices} = Cu.import("resource://gre/modules/devtools/Devices.jsm");
-const {when: unload} = require("sdk/system/unload");
+const { defineLazyGetter } = devtoolsRequire("devtools/toolkit/DevToolsUtils");
 
-let promise;
-try {
-  promise = Cu.import("resource://gre/modules/commonjs/promise/core.js").Promise;
-} catch (e) {
-  promise = Cu.import("resource://gre/modules/commonjs/sdk/core/promise.js").Promise;
-}
+const events = require("sdk/event/core");
+const { when: unload } = require("sdk/system/unload");
+const { Devices } = Cu.import("resource://gre/modules/devtools/Devices.jsm");
+const { gDevToolsBrowser } =
+  Cu.import("resource:///modules/devtools/gDevTools.jsm");
+defineLazyGetter(this, "adb", () => {
+  return require("./adb");
+});
+defineLazyGetter(this, "fastboot", () => {
+  return require("./fastboot");
+});
+defineLazyGetter(this, "Device", () => {
+  return require("./device");
+});
 
-// As of Firefox 36, WebIDE exposes an API to register new runtimes.
-try {
+// Set this right away on startup
+Devices.helperAddonInstalled = true;
+unload(() => Devices.helperAddonInstalled = false);
+
+function onADBStart() {
+  // As of Firefox 36, WebIDE exposes an API to register new runtimes.
   let Runtimes = devtoolsRequire("devtools/webide/runtimes");
   if (Runtimes && Runtimes.RuntimeScanners) {
     let scanner = require("./scanner");
     scanner.register();
   }
-} catch (e) {}
-
-Devices.helperAddonInstalled = true;
-exports.shutdown = function() {
-  Devices.helperAddonInstalled = false;
-  adb.stop(true);
-};
-
-adb.start().then(function () {
-  adb.trackDevices();
-});
 
-/**
- * A Device instance is created and registered with the Devices module whenever
- * ADB notices a new device is connected.
- *
- * Any changes here should be examined carefully for backwards compatibility
- * issues.  Other add-ons, like the Tools Adapter, make use of these low-level
- * device objects too.
- */
-function Device(id) {
-  this.id = id;
+  adb.start().then(function() {
+    adb.trackDevices();
+  });
 }
+Devices.on("adb-start-polling", onADBStart);
```

Comment:
Having to depend on FF41+ is quite painful as we are going to update the addon for everyone, not just FF41+.
(I think that the addon manager will try to update, even if the manifest states that it's not compatible...)

It would be great if you could listen for webide promise on FF41- (and ensure that ADBStart is executed only once).
The b2d addon is going to rely on adb-start-polling no matter what is the FF version.